1927 revisit all evens that make upsert

devyetii opened the pull request:
## Fixes
Use UPDATE Whenever upsert is not needed or incomplete.
## Changes
- Use the updater whenever `UpdateAll: true` is not set in the upsert statement, excluding when the function name states `addOrOverwrite*`

## Need to be mentioned in CHANGELOG.md?

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

devyetii committed:
Add columnize util method

devyetii committed:
use columnizer in update validator

devyetii committed:
Fix allocation upserts

devyetii committed:
Add updater extra conditions

devyetii committed:
Fix challenges and terms upsert

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
Fix updater unit test

devyetii committed:
Add extraCondition updater unit testing

devyetii committed:
Complete extraCondition unit testing

devyetii committed:
Remove set statement of extra conditions

devyetii committed:
Fix lint

devyetii committed:
Add extraColumns to unnests

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
Breakdown updater functions into reusable parts

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/updater_test.go:
```diff
@@ -30,15 +36,15 @@ func TestUpdateBuilder_build(t *testing.T) {
 				ids:     interface{}([]string{"1"}),
 				updates: []update{{val: []string{"c1"}, key: "column"}},
 			},
-			want: "UPDATE table SET column = t.column FROM (SELECT unnest(?::text[]) AS id, unnest(?::text[]) AS column) AS t WHERE table.id = t.id",
```
should not be changed

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/updater_test.go:
```diff
@@ -30,15 +36,15 @@ func TestUpdateBuilder_build(t *testing.T) {
 				ids:     []int{1},
 				updates: []update{{val: []string{"c1"}, key: "column"}},
 			},
-			want: "UPDATE table SET column = t.column FROM (SELECT unnest(?::bigint[]) AS id, unnest(?::text[]) AS column) AS t WHERE table.id = t.id",
```
should not be changed

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/updater_test.go:
```diff
@@ -51,7 +57,7 @@ func TestUpdateBuilder_build(t *testing.T) {
 					{val: []string{"c41", "c42", "c43"}, key: "column4"},
 				},
 			},
-			want: "UPDATE table SET column1 = t.column1, column2 = t.column2, column3 = t.column3, column4 = t.column4 FROM (SELECT unnest(?::text[]) AS id, unnest(?::text[]) AS column1, unnest(?::text[]) AS column2, unnest(?::text[]) AS column3, unnest(?::text[]) AS column4) AS t WHERE table.id = t.id",
```
should not be changed

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/updater.go:
```diff
@@ -59,6 +63,39 @@ func CreateBuilder(table string, idColumn string, idValues interface{}) (b *Upda
+}
+
+// AddCondition Add ANDed condition to the update query 
+func (b *UpdateBuilder) AddCondition(column string, values interface{}) *UpdateBuilder {
```
what is this special conditions, if it is a part of composite key we should add it to the ID logic, not as additional clause

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/updater_test.go:
```diff
@@ -90,7 +96,21 @@ func TestUpdateBuilder_build(t *testing.T) {
+					{key: "id3", val: []string{"1_3", "2_3", "3_3"}},
+				},
+			},
+			want: "UPDATE table SET column1 = t.column1 FROM (SELECT unnest(?::text[]) AS id, unnest(?::text[]) AS id2, unnest(?::text[]) AS id3, unnest(?::text[]) AS column1) AS t WHERE table.id = t.id AND table.id2 = t.id2 AND table.id3 = t.id3 ",
```
it's not additional condition, but part of ID, additional condition would be 
```sql
WHERE table.temperature > 17
```
but not join clause

dabasov commented on code/go/0chain.net/core/util/columnizer.go:
```diff
@@ -0,0 +1,41 @@
+
+		for fidx := 0; fidx < t.NumField(); fidx++ {
+			f := t.Field(fidx)
+			fkey := f.Tag.Get("json")
```
not sure json is a way to go. we should consider raw names and conversion to snake case, together with gorm overwriting. 
field name can be overwritten with gorm tag, after goose migrator is merged it is simply written in the db and there is no info in golang code of it's name, so mapping should be done somehow extensively, we still can go on work with gorm column names, but should add it here as well.

dabasov replied:
@devyetii why do you still consider json tags?

devyetii replied:
I didn't disregard it totally, instead supported two more naming methods, which are gorm and snake-casing with precedence order of JSON tag then gorm then snake-casing, just in case a field needs to be named specially not following the snake-casing rules. Should this totally be disregarded ?

dabasov replied:
I think yes, since gorm disregards it by default

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
Resolve comments on updater

devyetii committed:
Resolve comment in columnizer

devyetii committed:
Merge branch '1927-revisit-all-evens-that-make-upsert' of github.com:0chain/0chain into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
fix columnizer

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/allocation.go:
```diff
@@ -129,6 +129,28 @@ func (edb *EventDb) updateAllocations(allocs []Allocation) error {
 		"latest_closed_challenge_txn",
```
why do you have this list if you already make `util.Columnize(allocs)`

devyetii replied:
This list was there before I added the columnizer. The columnizer returns all the possible fields in the struct, where only these fields were originally used in the update. I just didn't like to change the behaviour.

dabasov replied:
what's reason to use Columnize then? I think it is to collect all the fields that are not ignored

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/allocation.go:
```diff
@@ -129,6 +129,28 @@ func (edb *EventDb) updateAllocations(allocs []Allocation) error {
+			continue
+		}
+
+		fieldList, ok := columns[fieldKey]
```
simply use columns you got from Columnize

devyetii replied:
Same as above

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/allocation.go:
```diff
@@ -129,6 +129,28 @@ func (edb *EventDb) updateAllocations(allocs []Allocation) error {
+		if !ok {
+			return corecommon.NewErrorf("update_allocation", "required field %v for update is not found in provided data", fieldKey)
+		}
+		updater = updater.AddUpdate(fieldKey, fieldList)
```
do not run updater in loop, you should pass here all collected arrays of values

dabasov replied:
you should get an array of values for every field actually

devyetii replied:
That's what Columnize does, it returns an array of values for each field. I'm just simply adding them to the updater instance. Running the updater is down in line 163

dabasov replied:
ah good

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/allocation.go:
```diff
@@ -191,19 +242,47 @@ func withAllocStatsMerged() eventMergeMiddleware {
+		AddUpdate("num_writes", numWritesList, "allocations.num_writes + t.num_writes").
+		AddUpdate("moved_to_challenge", movedToChallengeList, "allocations.moved_to_challenge + t.moved_to_challenge").
+		AddUpdate("moved_back", movedBackList, "allocations.moved_back + t.moved_back").
+		AddUpdate("write_pool", writePoolList, "allocations.write_pool - t.write_pool").Exec(edb).Error
```
we had 
```
allocations.write_pool - excluded.moved_to_challenge + excluded.moved_back
```
before, why did it change?

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/updater.go:
```diff
@@ -43,14 +45,17 @@ type UpdateBuilder struct {
 }
 
+// AddIdPart Add new part of a composite id key
+func (b *UpdateBuilder) AddIdPart(columnName string, values interface{}) *UpdateBuilder {
```
`AddCompositeId` will be better I think

devyetii replied:
Done

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/updater.go:
```diff
@@ -83,27 +119,27 @@ func (b *UpdateBuilder) AddUpdate(column string, values interface{}, expr ...str
-	for _, s := range b.sets {
-		sets = sets + s
+func (b *UpdateBuilder) addWhereCondition(left, operator, right string) *UpdateBuilder {
+	if b.where != "" {
```
please remove this if to more clear way of creating string, the first where will be added by JOIN clause I believe

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/validator.go:
```diff
@@ -109,10 +111,34 @@ func (edb *EventDb) updateValidators(validators []Validator) error {
+		if !ok {
+			return common.NewErrorf("update_validators", "required field %v for update is not found in provided data", fieldKey)
+		}
+		updater = updater.AddUpdate(fieldKey, fieldList)
```
same as for allocation

devyetii replied:
That would be easier for me, but, as I mentioned above, I just didn't like to add any fields to the update that wasn't there before.

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/updater_test.go:
```diff
@@ -92,13 +98,32 @@ func TestUpdateBuilder_build(t *testing.T) {
 			},
 			want: "UPDATE table SET column1 = column1 + t.column1, column2 = column2 - t.column2, column3 = t.column3, column4 = column4 * t.column4 FROM (SELECT unnest(?::text[]) AS id, unnest(?::text[]) AS column1, unnest(?::bigint[]) AS column2, unnest(?::bytea[]) AS column3, unnest(?::decimal[]) AS column4) AS t WHERE table.id = t.id",
 		},
+		{
```
add tests for `addWhereCondition`

devyetii committed:
Fix write pool update

devyetii committed:
AddIdParts to AddCompositeId

devyetii committed:
Simplify where condition

devyetii committed:
Revert "Simplify where condition"

This reverts commit 99c2cc61ce0e4e9b3a6e725874a3117eb5e6ece4.

devyetii committed:
separate where condition func for id

devyetii committed:
add unit test for AddCondition

devyetii committed:
fix unit tests

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
disregard json field in columnizer

devyetii committed:
Merge branch '1927-revisit-all-evens-that-make-upsert' of github.com:0chain/0chain into 1927-revisit-all-evens-that-make-upsert

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/allocation.go:
```diff
@@ -258,29 +337,45 @@ func withAllocChallengesMerged() eventMergeMiddleware {
 }
 
 func (edb *EventDb) updateAllocationChallenges(allocs []Allocation) error {
```
leet's have a test with db for this method

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/allocation.go:
```diff
@@ -258,29 +337,45 @@ func withAllocChallengesMerged() eventMergeMiddleware {
+		AddUpdate("failed_challenges", failedChallengeList, "allocations.failed_challenges + t.failed_challenges").Exec(edb).Error
 }
 
 func (edb *EventDb) addChallengesToAllocations(allocs []Allocation) error {
```
let's have a test with db for this method

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/allocation_blobber_term.go:
```diff
@@ -75,7 +75,30 @@ func (edb *EventDb) deleteAllocationBlobberTerms(terms []AllocationBlobberTerm)
 }
 
 func (edb *EventDb) updateAllocationBlobberTerms(terms []AllocationBlobberTerm) error {
```
let's have a test with db for this method

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/challenge.go:
```diff
@@ -78,10 +78,21 @@ func (edb *EventDb) addChallenges(chlgs []Challenge) error {
 }
 
 func (edb *EventDb) updateChallenges(chs []Challenge) error {
```
let's have a test with db for this method

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/allocation.go:
```diff
@@ -159,17 +178,49 @@ func (edb *EventDb) updateAllocationStakes(allocs []Allocation) error {
 		}
 	}()
```
let's have a test with db for this method

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/allocation.go:
```diff
@@ -191,19 +242,47 @@ func withAllocStatsMerged() eventMergeMiddleware {
 }
 
 func (edb *EventDb) updateAllocationsStats(allocs []Allocation) error {
```
let's have a test with db for this method

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
Add unit tests to allocations

devyetii committed:
Fix unit test

devyetii committed:
fix unit test

devyetii committed:
fix unit tests

devyetii committed:
add allocation blobber terms unit test

devyetii committed:
fix unit tests

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
fix unit tests

devyetii committed:
better casting of allocation blobber terms max offer duration

devyetii committed:
fix unit test

devyetii committed:
add challenge unit test

devyetii committed:
add boolean type to updater

devyetii committed:
fix unit tests

peterlimg commented on code/go/0chain.net/core/util/columnizer.go:
```diff
@@ -0,0 +1,61 @@
+		for fidx := 0; fidx < t.NumField(); fidx++ {
+			f := t.Field(fidx)
+			
+			// Resolve field key in this order: gorm.column Tag > snake-casing FieldName
```
We need tests to ensure the field keys are resolved as expected. So let's add tests to cover all the cases.

peterlimg commented on code/go/0chain.net/core/util/columnizer.go:
```diff
@@ -0,0 +1,61 @@
+    return strings.ToLower(snake)
+}
+
+func Columnize[T any](objects []T) (map[string][]interface{}, error) {
```
We shall avoid putting code in `util` package when possible, we can put it in `dbs/event`, where the `updater.go` located. Or have its own package in `dbs`.

peterlimg commented on code/go/0chain.net/smartcontract/dbs/event/allocation.go:
```diff
@@ -162,17 +181,49 @@ func (edb *EventDb) updateAllocationStakes(allocs []Allocation) error {
+		movedBackList		  	[]int64
+		movedToValidatorsList 	[]int64
+		coinValue				int64
+		err						error
```
Run `goimports` to format the variables above.

peterlimg reviewed (changes requested):
Good job 👍 , a few minor change requests

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
columnizer in dbs/events

devyetii committed:
fix lint issues in dbs/event/allocation.go

devyetii committed:
columnizer unit test

fix unit test

fix unit testing

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

devyetii commented:
@peterlimg all comments resolved, pls check. Kindly note sys tests are still failing and I'll be resolving them tomorrow.

peterlimg commented on code/go/0chain.net/core/util/columnizer.go:
```diff
@@ -0,0 +1,61 @@
+package util
```
Please remove this file as it's in event package now.

peterlimg commented:
@devyetii can you rebase those `fix unit test` commits to keep the logs clean, thanks.

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
delete old columnizer

devyetii commented:
@peterlimg All resolved. Kindly check.

peterlimg reviewed (approved):
LGTM. Please fix the system tests, seems the changes has something broken that caused the API tests failed.

devyetii committed:
allocation blobber terms should use add/overwrite

devyetii committed:
fix unit test

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | fix/gosdk-to-staging  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | 1927-revisit-all-evens-that-make-upsert  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

devyetii committed:
warning is enough when a field not found in request

devyetii committed:
fix validator

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
fix duplicate

devyetii committed:
Add update allocation unit testing

devyetii commented:
@dabasov @peterlimg found a bug in the columnizer+updater. The `[]interface{}`'s returned from the columnizer when passed into `reflect.TypeOf` return `[]interface{}` so not found in `typeToSQL` so all fields are casted to `text` sql type. Working on that + added some unit tests.

devyetii committed:
fix updater couldn't resolve type of []interface{}

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

devyetii committed:
Add support of currency.Coin to updater

devyetii committed:
Add unit testing to validator

devyetii committed:
use straight array definitions instead of columnizer for validator

devyetii committed:
fix validator test

devyetii commented:
@dabasov @peterlimg One more limitation of the current version of the columnizer is that it fails in converting nested structs, for example any struct that has `Provider` as part of its definition. To speed up merging this PR I'll fix it in another PR and add more unit tests.

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | 1927-revisit-all-evens-that-make-upsert  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

devyetii committed:
Enhance columnizer

dabasov reviewed (approved):
lgtm

devyetii committed:
Revert "use straight array definitions instead of columnizer for validator"

This reverts commit 17d8c5f3c301c736f855cd011c28c67720d80464.

devyetii committed:
more columnizer enhances + fix tests

fix unit tests

fix unit tests

cleaning columnizer

fix unit test

Revert "fix unit test"

This reverts commit efd7c5281bef220c6003144a682c297772f66b55.

enhance columnizer + fix tests

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | 1927-revisit-all-evens-that-make-upsert  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

devyetii commented:
@peterlimg @dabasov kindly review the latest changes

devyetii committed:
Merge branch 'staging' into 1927-revisit-all-evens-that-make-upsert

peterlimg commented:
LGTM
